### PR TITLE
add DistributionScheduler contract

### PR DESCRIPTION
**Pull request by TomAFrench**
This contract aims to allow partners to schedule distributions of their tokens on gauges without needing to fire off a transaction at the same time every week, instead they can queue up multiple weeks worth of rewards and leave the triggering of the distribution to bots like Defender.

Anyone may provide tokens to the `DistributionScheduler` along with a timestamp at which it is to be sent to the gauge, this is then stored in a linked list of all the other distributions of this token to a particular gauge. We can then just iterate through this list up to the current timestamp in order to start any distributions which are scheduled to do so.

We require that distributions start at Thurs 00:00 UTC for consistency with BAL rewards and also to prevent DOS from having many 1 wei distributions which each need to be processed before the next meaningful distribution.

### CODE REVIEW

File: pkg/liquidity-mining/contracts/test/MockRewardTokenDistributor.sol [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+pragma experimental ABIEncoderV2;
+
+import "@balancer-labs/v2-solidity-utils/contracts/openzeppelin/IERC20.sol";
+
+// solhint-disable func-name-mixedcase, var-name-mixedcase, not-rely-on-time
+
+/**
+ * @dev This contract is designed to mock LiquidityGaugeV5's interface for distributing external tokens.
+ */
+contract MockRewardTokenDistributor {
```
Comment (nventuro) [inline pkg/liquidity-mining/contracts/test/MockRewardTokenDistributor.sol:25]: Let's also inherit from the interface, given that's what we're mocking and interacting with in the tests.

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+    });
+
+    context('when DistributionScheduler is the distributor for reward token', () => {
+      sharedBeforeEach('set reward token distributor to DistributionScheduler', async () => {
+        await rewardTokenDistributor.add_reward(rewardToken.address, distributionScheduler.address);
+      });
+
+      context('when distribution is scheduled in the past', () => {
+        it('reverts', async () => {
+          await expect(scheduleDistribution(1, 0)).to.be.revertedWith(
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:111]:
```suggestion
          await expect(scheduleDistribution(1, (await currentTimestamp()).sub(1))).to.be.revertedWith(
```

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+    context("when distribution starts at a timestamp which doesn't fit in a uint32", () => {
+      it('reverts', async () => {
+        await expect(scheduleDistribution(1, MAX_UINT32.add(1))).to.be.revertedWith('Reward timestamp overflow');
+      });
+    });
+
+    context('when reward token does not exist on gauge', () => {
+      it('reverts', async () => {
+        await expect(scheduleDistribution(1, 0)).to.be.revertedWith('Reward token does not exist on gauge');
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:88]: I'd consider changing these hardcoded 1's to e.g. `const amount = 42`. This is specially useful in later tests.

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+
+            context('when a reward is being added after the last existing reward', () => {
+              let insertedTime: BigNumber;
+
+              sharedBeforeEach('set insertedTime', async () => {
+                insertedTime = startTime.add(20 * WEEK);
+              });
+
+              it('updates the previous node to point at the new node', async () => {
+                const prevNodeKey = insertedTime.sub(2 * WEEK);
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:175]: This is a bit magical as it relies on how nodes were constructed in the `for` loop. I'd consider replacing this with a search to find the last node key.

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+              for (let i = 0; i < 10; i++) {
+                await scheduleDistribution(100, startTime.add(i * 2 * WEEK));
+              }
+            });
+
+            context('when a reward is being added after the last existing reward', () => {
+              let insertedTime: BigNumber;
+
+              sharedBeforeEach('set insertedTime', async () => {
+                insertedTime = startTime.add(20 * WEEK);
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:171]:
```suggestion
                insertedTime = startTime.add(20 * WEEK);
```

Same here, we're _right_ at the edge of constructed values. I'd consider doing something like `9999 * WEEK`.

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+
+              it('writes the expected node', async () => {
+                const oldRewardNode = await getRewardNode(insertedTime);
+                expect(oldRewardNode.amount).to.be.eq(0);
+
+                await scheduleDistribution(1, insertedTime);
+
+                const newRewardNode = await getRewardNode(insertedTime);
+                expect(newRewardNode.amount).to.be.eq(1);
+                expect(newRewardNode.nextNodeKey).to.be.eq(insertedTime.add(WEEK));
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:219]:
```suggestion
                expect(newRewardNode.nextNodeKey).to.be.eq(oldRewardNode.nextNodeKey);
```

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+            });
+
+            it('writes the expected node', async () => {
+              const oldRewardNode = await getRewardNode(startTime);
+              expect(oldRewardNode.amount).to.be.eq(0);
+
+              await scheduleDistribution(1, startTime);
+
+              const newRewardNode = await getRewardNode(startTime);
+              expect(newRewardNode.amount).to.be.eq(1);
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:156]:
```suggestion
              expect(newRewardNode.amount).to.be.eq(1);
              expect(newRewardNode.nextNodeKey)to.be.eq(NULL);
```

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+              });
+
+              it('writes the expected node', async () => {
+                const oldRewardNode = await getRewardNode(insertedTime);
+                expect(oldRewardNode.amount).to.be.eq(100);
+
+                await scheduleDistribution(1, insertedTime);
+
+                const newRewardNode = await getRewardNode(insertedTime);
+                expect(newRewardNode.amount).to.be.eq(101);
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:246]: If we go with `amount` I'd rewrite this to e.g. `oldRewardNode.amount + 1` (or even `+ newAmount`), which is a bit more readable.

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 9]
```diff
@@ -0,0 +1,394 @@
...
+                expect(newRewardNode.nextNodeKey).to.be.eq(oldRewardNode.nextNodeKey);
+              });
+
+              context('when providing more tokens would cause an overflow on the node', () => {
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:250]: I wrote a note on my notebook to check if this test case existed while reviewing the code. Well done :grin:

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+                it('reverts', async () => {
+                  const { amount: existingRewardAmount } = await getRewardNode(insertedTime);
+                  const rewardAmount = MAX_UINT224.sub(1);
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:253]:
```suggestion
                  const rewardAmount = MAX_UINT224.sub(existingRewardAmount).add(1);
```

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 11]
```diff
@@ -0,0 +1,394 @@
...
+      return tx.wait();
+    }
+
+    sharedBeforeEach('set reward token distributor to DistributionScheduler', async () => {
+      await rewardTokenDistributor.add_reward(rewardToken.address, distributionScheduler.address);
+      startTime = roundUpTimestamp(await currentTimestamp());
+      for (let i = 0; i < 10; i++) {
+        await scheduleDistribution(100, startTime.add(i * WEEK));
+      }
+      await advanceToTimestamp(startTime.add(1));
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:287]: I think we don't need this `add(1)`?

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 12]
```diff
@@ -0,0 +1,394 @@
...
+
+        await startDistributionForToken();
+
+        expect(await getNextNodeKey(HEAD)).to.be.eq(startTime.add(WEEK));
+      });
+    });
+
+    context('when the last scheduled distribution is processed', () => {
+      sharedBeforeEach('advance time past the last scheduled distribution', async () => {
+        await advanceToTimestamp(startTime.add(10 * WEEK));
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:328]: Same comment re. scheduled construction in the for loop and this magical `10`

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 13]
```diff
@@ -0,0 +1,394 @@
...
+
+    sharedBeforeEach('deploy Token', async () => {
+      rewardToken2 = await Token.create('REWARD2');
+      await rewardToken2.mint(caller);
+      await rewardToken2.approve(distributionScheduler, MAX_UINT256, { from: caller });
+    });
+
+    sharedBeforeEach('initialize scheduled distributions', async () => {
+      await rewardTokenDistributor.add_reward(rewardToken.address, distributionScheduler.address);
+      await rewardTokenDistributor.add_reward(rewardToken2.address, distributionScheduler.address);
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:353]: Let's add a third reward with another distributor and check that it is correctly skipped

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 14]
```diff
@@ -0,0 +1,394 @@
...
+      const tx = await distributionScheduler.startDistributions(rewardTokenDistributor.address);
+      const receipt = await tx.wait();
+
+      expectEvent.inIndirectReceipt(
+        receipt,
+        rewardToken.instance.interface,
+        'Transfer',
+        {
+          from: distributionScheduler.address,
+          to: rewardTokenDistributor.address,
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:377]: We could also add `amount` here, populated with `getPendingRewards`

**Review by nventuro — COMMENTED** [review]
Lovely :grin: You've earned comp-sci fun permission!

File: pkg/liquidity-mining/contracts/admin/DistributionScheduler.sol [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+        uint256 startTime
+    ) external {
+        require(amount > 0, "Must provide non-zero number of tokens");
+
+        // Ensure that values won't overflow when put into storage.
+        require(amount <= type(uint224).max, "Reward amount overflow");
+        require(startTime <= type(uint32).max, "Reward timestamp overflow");
+
+        // Ensure that a user doesn't add a reward token which becomes locked on scheduler
+        address rewardDistributor = gauge.reward_data(token).distributor;
```
[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/liquidity-mining/contracts/admin/DistributionScheduler.sol:95]: I'm not a huge fan of this as I think it's overly restrictive. We'll need to special case any gauges which receive rewards but don't exist on the gauge controller.

We've had situations recently where partners (Gnosis in this case) want to get rewards started the same day and we obviously can't add gauges to the controller on that sort of turnaround. They then need to do hacky workarounds passing the distributor role to various addresses in order to get the first rewards out and then get it set up properly once the gauge is on the controller.

tl;dr I see more potential for friction on the ops side from adding this than risk from someone yeeting tokens into the void not checking the address they're sending it to.

File: pkg/liquidity-mining/contracts/admin/DistributionScheduler.sol [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+     */
+    function _insertReward(
+        mapping(uint32 => RewardNode) storage rewardsList,
+        uint32 insertedNodeKey,
+        uint224 amount
+    ) private {
+        // We want to find two nodes which sit either side of the new node to be created so we can insert between them.
+
+        uint32 currentNodeKey = _HEAD;
+        uint32 nextNodeKey = rewardsList[currentNodeKey].nextTimestamp;
```
Comment (nventuro) [inline pkg/liquidity-mining/contracts/admin/DistributionScheduler.sol:185]: Are these intentionally not renamed?

File: pkg/liquidity-mining/test/DistributionScheduler.test.ts [hunk 17]
```diff
@@ -0,0 +1,424 @@
...
+          await expect(scheduleDistribution(amount, (await currentTimestamp()).sub(1))).to.be.revertedWith(
+            'Distribution can only be scheduled for the future'
+          );
+        });
+      });
+
+      context('when distribution is scheduled in the future', () => {
+        let startTime: BigNumber;
+
+        sharedBeforeEach('schedule some existing distributions', async () => {
```
Comment (nventuro) [inline pkg/liquidity-mining/test/DistributionScheduler.test.ts:122]:
```suggestion
        sharedBeforeEach('set the first valid timestamp to schedule rewards', async () => {
```

File: pkg/liquidity-mining/contracts/admin/DistributionScheduler.sol (hunk 15 above)
[the remark this answers is not in the document]
Comment (nventuro) [reply pkg/liquidity-mining/contracts/admin/DistributionScheduler.sol:97]: That's a very good point, thank.

In such a scenario, the LM committee would still need to manually add the reward tokens though, right?

[the remark this answers is not in the document]
Comment (TomAFrench) [reply pkg/liquidity-mining/contracts/admin/DistributionScheduler.sol:97]: Yeah, they'd need to add the token and set the scheduler as the distributor but that would be one tx rather than several.

**Review by nventuro — APPROVED** [review]
